File: server/src/main/java/org/hkurh/doky/documents/DocumentServiceImpl.java
package org.hkurh.doky.documents;

import org.hkurh.doky.users.UserService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static java.lang.String.format;

@Service
public class DocumentServiceImpl implements DocumentService {

    private static final Logger LOG = LoggerFactory.getLogger(DocumentServiceImpl.class);

    private final DocumentEntityRepository documentEntityRepository;
    private final UserService userService;

    public DocumentServiceImpl(DocumentEntityRepository documentEntityRepository, UserService userService) {
        this.documentEntityRepository = documentEntityRepository;
        this.userService = userService;
    }

    @Override
    public DocumentEntity create(@NonNull String name, String description) {
        var document = new DocumentEntity();
        document.setName(name);
        document.setDescription(description);
        var currentUser = userService.getCurrentUser();
        document.setCreator(currentUser);
        LOG.debug(format("Created new Document [%s] by User [%s]", document.getId(), currentUser.getId()));
        return documentEntityRepository.save(document);
    }

    @Override
    public Optional<DocumentEntity> find(@NonNull String id) {
        var documentId = Long.parseLong(id);
        var currentUser = userService.getCurrentUser();
        return documentEntityRepository.findByIdAndCreatorId(documentId, currentUser.getId());
    }

    @Override
    public @NonNull List<DocumentEntity> find() {
        var currentUser = userService.getCurrentUser();
        return new ArrayList<>(documentEntityRepository.findByCreatorId(currentUser.getId()));
    }

    @Override
    public void save(@NonNull DocumentEntity document) {
        documentEntityRepository.save(document);
    }
}

File: server/src/main/java/org/hkurh/doky/documents/DocumentServiceImpl.kt
package org.hkurh.doky.documents

import org.hkurh.doky.users.UserService
import org.slf4j.LoggerFactory
import org.springframework.lang.NonNull
import org.springframework.stereotype.Service
import java.util.*

@Service
class DocumentServiceImpl(private val documentEntityRepository: DocumentEntityRepository, private val userService: UserService) : DocumentService {
    override fun create(@NonNull name: String?, description: String?): DocumentEntity? {
        val document = DocumentEntity()
        document.name = name
        document.description = description
        val currentUser = userService.currentUser
        document.creator = currentUser
        LOG.debug(String.format("Created new Document [%s] by User [%s]", document.id, currentUser!!.id))
        return documentEntityRepository.save(document)
    }

    override fun find(@NonNull id: String?): Optional<DocumentEntity?>? {
        val documentId = id!!.toLong()
        val currentUser = userService.currentUser
        return documentEntityRepository.findByIdAndCreatorId(documentId, currentUser!!.id)
    }

    @NonNull
    override fun find(): List<DocumentEntity?>? {
        val currentUser = userService.currentUser
        return documentEntityRepository.findByCreatorId(currentUser!!.id)?.let { ArrayList(it) }
    }

    override fun save(@NonNull document: DocumentEntity?) {
        documentEntityRepository.save(document!!)
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(DocumentServiceImpl::class.java)
    }
}
